examples_tests/49.NewAPITest/main.cpp

```diff
@@ -0,0 +1,330 @@
+		auto vs = device->createGPUSpecializedShader(vs_unspec.get(), vsinfo);
+		asset::ISpecializedShader::SInfo fsinfo(nullptr, nullptr, "main", asset::ISpecializedShader::ESS_FRAGMENT, "fs");
+		auto fs = device->createGPUSpecializedShader(fs_unspec.get(), fsinfo);
+
+		video::IGPUSpecializedShader* shaders[2]{ vs.get(), fs.get() };
+
+		asset::SVertexInputParams vtxinput;
+		vtxinput.attributes[0].binding = 0;
+		vtxinput.attributes[0].format = asset::EF_R32G32_SFLOAT;
+		vtxinput.attributes[0].relativeOffset = offsetof(SVertex, pos);
+
+		vtxinput.attributes[1].binding = 0;
+		vtxinput.attributes[1].format = asset::EF_R32G32B32_SFLOAT;
+		vtxinput.attributes[1].relativeOffset = offsetof(SVertex, color);
+
+		vtxinput.bindings[0].inputRate = asset::EVIR_PER_VERTEX;
+		vtxinput.bindings[0].stride = sizeof(SVertex);
+
+		vtxinput.enabledAttribFlags = 0b0011;
+		vtxinput.enabledBindingFlags = 0b0001;
+		
+		asset::SRasterizationParams raster;
+		raster.depthTestEnable = 0;
+		raster.depthWriteEnable = 0;
+		raster.faceCullingMode = asset::EFCM_NONE;
+		
+		asset::SPrimitiveAssemblyParams primitive;
+		primitive.primitiveType = asset::EPT_TRIANGLE_LIST;
+
+		asset::SBlendParams blend;
+
+		rpindependent_pipeline = device->createGPURenderpassIndependentPipeline(nullptr, core::smart_refctd_ptr(layout), shaders, shaders+2, vtxinput, blend, primitive, raster);
+		assert(rpindependent_pipeline);
+	}
+
+	core::smart_refctd_ptr<video::IGPUGraphicsPipeline> pipeline;
+	{
+		video::IGPUGraphicsPipeline::SCreationParams gp_params;
+		gp_params.rasterizationSamplesHint = asset::IImage::ESCF_1_BIT;
+		gp_params.renderpass = renderpass;
+		gp_params.renderpassIndependent = rpindependent_pipeline;
+		gp_params.subpassIx = 0u;
+
+		pipeline = device->createGPUGraphicsPipeline(nullptr, std::move(gp_params));
+	}
+
+	core::smart_refctd_ptr<video::IGPUBuffer> buffer;
+	{
+		const SVertex vertices[3]{
+			{{0.0f, -0.5f}, {1.0f, 0.0f, 0.0f}},
+			{{0.5f, 0.5f}, {0.0f, 1.0f, 0.0f}},
+			{{-0.5f, 0.5f}, {0.0f, 0.0f, 1.0f}}
+		};
+
+		video::IDriverMemoryBacked::SDriverMemoryRequirements mreq;
+		
+		buffer = device->createDeviceLocalGPUBufferOnDedMem(sizeof(vertices));
+		assert(buffer);
+
+		core::smart_refctd_ptr<video::IGPUCommandBuffer> cb;
+		device->createCommandBuffers(cmdpool.get(), video::IGPUCommandBuffer::EL_PRIMARY, 1u, &cb);
+		assert(cb);
+
+		cb->begin(video::IGPUCommandBuffer::EU_ONE_TIME_SUBMIT_BIT);
+		cb->updateBuffer(buffer.get(), 0u, sizeof(vertices), vertices);
+		cb->end();
+		
+		video::IGPUQueue::SSubmitInfo info;
+		auto* cb_ = dynamic_cast<video::IGPUPrimaryCommandBuffer*>(cb.get());
+		info.commandBufferCount = 1u;
+		info.commandBuffers = &cb_;
+		info.pSignalSemaphores = nullptr;
+		info.signalSemaphoreCount = 0u;
+		info.pWaitSemaphores = nullptr;
+		info.waitSemaphoreCount = 0u;
+		info.pWaitDstStageMask = nullptr;
+		queue->submit(1u, &info, nullptr);
```

Comment:
I understand that the drawing cmd buffers wll be submitted to the same queue later, but you need a memory barrier (under Vulkan).

So a pipeline barrier should in the cmd buffer at the end with SRC stage being TRANSFER, and DST being Vertex Input, with a memory barrier over the buffer you've just updated.

Reply:
yea, right

Reply:
@achalpandeyy synchro chad, please verify

Reply:
nevermind, fence signal is a full availability op, while submit is a full visibility op